transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java

```diff
@@ -333,70 +363,94 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
      * @param client instance of StftClient
      * @param fileAge How old the file should be to be considered "old" and not
      *            changed
-     * @param sizeCheckDelayMs Wait time (in ms) between size-checks to determine if
-     *            a file is ready to be processed.
      * @return true if the file has changed
      * @throws Exception Error
      */
-    protected boolean canProcessFile(String fileName, SftpClient client, long fileAge, long sizeCheckDelayMs)
-        throws Exception
+    protected boolean canProcessFile(String fileName, SftpClient client, long fileAge, List<String> stableFiles)
+            throws Exception
     {
-        if (fileAge > 0 && !isOldFile(fileName, client, fileAge))
+
+        if (fileAge > 0 && stableFiles != null)
         {
-            return false;
+            return isOldFile(fileName, client, fileAge) && stableFiles.contains(fileName);
         }
-
-        if (sizeCheckDelayMs > 0 && isSizeModified(fileName, client, sizeCheckDelayMs))
+        else if (stableFiles == null)
         {
-            return false;
+            return isOldFile(fileName, client, fileAge);
+        }
+        else if (fileAge <= 0)
+        {
+            return stableFiles.contains(fileName);
         }
 
         return true;
     }
 
-    private boolean isSizeModified(String fileName, SftpClient client, long sizeCheckDelayMs) throws IOException, InterruptedException
+    List<String> getStableFiles(String[] fileNames, SftpClient client, long sizeCheckDelayMs) throws InterruptedException
     {
-        try
+        List <String> stableFiles = new ArrayList<>(fileNames.length);
+        Map<String, Long> sizes = new HashMap<>();
+
+        if (logger.isDebugEnabled())
         {
-            if (logger.isDebugEnabled())
+            logger.debug("Perform size check with a delay of: " + sizeCheckDelayMs + " ms.");
+        }
+
+        for (String fileName : fileNames)
```

Comment:
Extract this to a method "getFileTimeStamps" or similar, then refactor the code to do something like:

a = getFileTimeStamps
sleep
b = getFileTimeStamps
filter a after comparing against b